Add validate cluster command to clusterctl

wangzhen127 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
This PR adds validate cluster command to clusterctl. It does not have real implementations yet. This is part of #168 and there will be follow-up PRs. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

<!-- All reviews default to cc'ing the kube-deploy-reviewers github group. -->
@kubernetes/kube-deploy-reviewers

wangzhen127 commented:
/hold
Should rebase and merge after #202 is merged first.

karan commented:
I can't review this PR with 11k files. Can you take out the vendor changes and only have the core change in this PR please?

wangzhen127 commented:
The benefit of including the 1st commit (vendor changes) are:
1. It gives you the context and why this PR depends on #202. 
2. This PR itself is complete and can compile and run. This is important because in terms of dev workflow, I need to check this PR works.
3. You can still review this PR easily. If you click on the commits tab, you will see 2 commits. The 2nd commit contains the actual changes. There are only 2 files and should be easy to review.

karan commented on clusterctl/cmd/validate_cluster.go:
```diff
@@ -0,0 +1,50 @@
+}
+
+func init() {
+	validateClusterCmd.Flags().StringVarP(&vco.Name, "name", "n", "", "Cluster name")
```
Please see jessica's comment here: https://github.com/kubernetes-sigs/cluster-api/pull/119#discussion_r185310852

> Please leave all flags out of the skeleton. They can be filled in as code uses them. This ensures the the flags actually do something when they appear and also ensure they are defined in a way that works and is accurate to the behavior of the code.

wangzhen127 replied:
OK. I just removed it. Will add them in follow-up PRs.

wangzhen127 commented:
Since we decided to drop internalization in #202 for now, I updated this PR to remove those changes. I also updated the PR description. PTAL

/hold cancel

spew commented on clusterctl/cmd/validate.go:
```diff
@@ -0,0 +1,32 @@
+}
+
+func init() {
+	validateCmd.AddCommand(validateClusterCmd)
```
To reduce confusion, we should be consistent with who is responsible for adding a sub-command. Since the validate command adds itself to the RootCmd on line 31, then it seems like we should remove line 30 and have the ValidateClusterCmd add itself to the ValidateCmd.

wangzhen127 replied:
By the way, is there a repo wide style guideline?

spew replied:
As you pointed out there's inconsistency in the repo. We can fix that in a follow up PR. Let's follow the style laid out in the cobra examples [1] where a given command adds itself to its parent. So that means the create command should be added in create.go. And in this case, the validateClusterCommand would add itself to the validate command in validate_cluster.go.

[1] https://github.com/spf13/cobra

jessicaochen replied:
@spew - thanks for standardizing
+1 for following the standardized pattern.

wangzhen127 replied:
Done

jessicaochen commented on clusterctl/cmd/validate.go:
```diff
@@ -0,0 +1,32 @@
+var validateCmd = &cobra.Command{
+	Use:   "validate",
+	Short: "Validate an API resource created by cluster API.",
+	Long:  "Validate an API resource created by cluster API.",
```
I would actually suggest more details on the validate cluster command instead of the validate command as I suspect validation will be cluster API resource-specific.

wangzhen127 replied:
I agree with Jessica.

And  I think it is very early and there is no real implementation yet. I intend to update the help message as I add more stuff in follow-up PRs.

wangzhen127 commented on clusterctl/cmd/validate_cluster.go:
```diff
@@ -0,0 +1,38 @@
+	},
+}
+
+func RunValidateCluster() error {
```
Done

wangzhen127 commented:
Updated the PR with fixes. PTAL

spew commented:
/lgtm

spew commented:
/approve

wangzhen127 commented:
I just noticed that testify was not in vendor/ from the failed test. So I added it in Gopkg.toml and ran dep ensure. Currently I put them as separate commits for easier review. I will squash after someone take a look.

spew commented:
So to date we have been avoiding testify in this repository and have been relying on writing out if statements. The reasoning comes from the golang FAQ:

https://golang.org/doc/faq#testing_framework
```
A related point is that testing frameworks tend to develop into mini-languages of their own, with conditionals and controls and printing mechanisms, but Go already has all those capabilities; why recreate them? We'd rather write tests in Go; it's one fewer language to learn and the approach keeps the tests straightforward and easy to understand.
```

I think we need a decision on this and make it a repository mandate or not. Maybe we can conduct this discussion in Slack or have an issue vote.

spew commented:
/ok-to-test

wangzhen127 commented:
OK. I updated the test to use if statement. PTAL

spew commented:
/lgtm
